### DISCUSSION TITLE
ssl handshake error on 9093 port number in strimzi kafka

### DISCUSSION THREAD

**Discussion by Cshanmukh80** [strimzi/strimzi-kafka-operator#4752] [Q&A]
**kafka.yaml** 
# Source: kafka/templates/kafka-persistent.yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: kafka
  namespace: default
  labels:
    helm.sh/chart: kafka-1.0
    app.kubernetes.io/name: kafka
    app.kubernetes.io/instance: kafka
    app.kubernetes.io/version: "1.0"
    app.kubernetes.io/managed-by: Helm
spec:
  kafka:
    version: 2.6.0
    replicas: 3
    template:
     pod:
       securityContext:
        runAsUser: 0
    listeners:
     - name: plain
       port: 9092
       type: internal
       tls: false
     - name: tls
       port: 9093
       type: internal
       tls: true
       authentication:
         type: tls
       configuration:
         brokerCertChainAndKey: 
           secretName: kafka-secret
           certificate: my_key_store_1615469373.crt
           key: my_store_1615469373.key
           #~ certificate: my_key_store.crt
           #~ key: my_store.key
     - name: external
       type: nodeport
       tls: false
       port: 9094
       configuration: 
         brokers:
         - broker: 0
           advertisedHost: 192.168.105.19
           advertisedPort: 32000
         - broker: 1
           advertisedHost: 192.168.105.19
           advertisedPort: 32001
         - broker: 2
           advertisedHost: 192.168.105.19
           advertisedPort: 32002
     - name: externalssl
       port: 9095
       type: nodeport
       tls: true
       authentication:
         type: tls
       configuration:
         brokerCertChainAndKey: 
           secretName: kafka-secret
           certificate: my_key_store_1615469373.crt
           key: my_store_1615469373.key
           #~ certificate: my_key_store.crt
           #~ key: my_store.key
         brokers:
         - broker: 0
           advertisedHost: 192.168.105.19
           advertisedPort: 32100
         - broker: 1
           advertisedHost: 192.168.105.19
           advertisedPort: 32101
         - broker: 2
           advertisedHost: 192.168.105.19
           advertisedPort: 32102
      #~ external:
        #~ type: nodeport
        #~ tls: false
        #~ overrides: 
          #~ brokers:
          #~ - broker: 0
            #~ advertisedHost: 192.168.105.19
            #~ advertisedPort: 32000
          #~ - broker: 1
            #~ advertisedHost: 192.168.105.19
            #~ advertisedPort: 32001
          #~ - broker: 2
            #~ advertisedHost: 192.168.105.19
            #~ advertisedPort: 32002
      #~ plain:
        #~ name: plain
        #~ port: 9092
        #~ type: internal
        #~ tls: false
      #~ tls: 
        #~ name: tls
        #~ port: 9093
        #~ type: internal
        #~ tls: true
        #~ authentication:
          #~ type: tls
        #~ configuration:
          #~ brokerCertChainAndKey: 
            #~ secretName: kafka-secret
            #~ certificate: my_key_store_1615469373.crt
            #~ key: my_store_1615469373.key
    config:
      offsets.topic.replication.factor: 1
      transaction.state.log.replication.factor: 1
      transaction.state.log.min.isr: 1
      log.message.format.version: "2.6"
      default.replication.factor: 3
      default.log.retention.ms: 2.88e+07
      num.partitions: 6 
      #~ ssl.cipher.suites: "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"
      ssl.enabled.protocols: "TLSv1.2"
      ssl.protocol: "TLSv1.2"
      ssl.endpoint.identification.algorithm: ""
      ssl.key.password: "kafka123"
      client.user: "user"
      client.password: "kafka123"
      inter.broker.listener.name: "PLAIN-9092"
    storage:
      type: ephemeral
  zookeeper:
    replicas: 3
    template:
     pod:
       securityContext:
        runAsUser: 0
    storage:
      type: ephemeral
    #~ config:
      #~ ssl.cipher.suites: "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"
      #~ ssl.enabled.protocols: "TLSv1.2"
      #~ ssl.protocol: "TLSv1.2"
      #~ ssl.endpoint.identification.algorithm: ""
  entityOperator:
    topicOperator: {}
    userOperator: {}

**clietn-ssl.properties**

security.protocol=SSL
ssl.truststore.type=PKCS12
ssl.truststore.location=/tmp/kafka/kafka.truststore.jks
ssl.truststore.password=kafka123
ssl.enabled.protocols=TLSv1.2
ssl.endpoint.identification.algorithm=

**i am getting error**

![Screenshot from 2021-04-14 11-57-07](https://user-images.githubusercontent.com/80450247/114663890-9afd4e80-9d18-11eb-9660-366f6797ef24.png)

can you help me for this issue

**Comment by scholzj** [comment]
It is quite hard to read the Kafka custom resource without proper formatting as I do not know the indentation, structure etc. So hard to say if it is correct or not or how does it look like.

**Reply by Cshanmukh80** [reply]
[kafka.zip](https://github.com/strimzi/strimzi-kafka-operator/files/6315443/kafka.zip)

sorry , please have a look in this file

**Reply by scholzj** [reply]
The YAML looks mostly ok. But in `.spec.kafka.config` you have quite a lot of options which will be ignored. For example:

```
      ssl.key.password: "kafka123"
      ssl.client.auth: "required"
      client.user: "user"
      client.password: "kafka123"
      inter.broker.listener.name: "PLAIN-9092"
```

Check the docs on what is allowed there: https://strimzi.io/docs/operators/latest/full/using.html#type-KafkaClusterSpec-schema-reference

----

Now, with your client, you are trying to connect to the listener on port 9093. However, I think you have several things wrong there:
* You should do this ideally from a separate pod and not from the Kafka pod
* You configuration file does not configure the keystore. But the listener has enabled TLS authentication. So you would not be able to authenticate. You either need to disable the authentication or create `KafkaUser` and use the certificate it creates for authentication
* I'm not sure where or how you got the `/tmp/kafka/kafka.truststore.jks` - are you sure it is the right one?

**Reply by Cshanmukh80** [reply]
sir, as you said, i have removed those configurations and also i have followed document which you have provided  but still it is giving ssl handshake issue

**Reply by scholzj** [reply]
Then maybe you should start by sharing your steps in detail:
* How did you created the truststore and keystore
* How does your configuration file look like right now
* etc.

You can also use the Java system property `javax.net.debug=ssl` to get additional TLS debug information from the client.

**Reply by Cshanmukh80** [reply]
[strimzi-kafka-ssl-folder.zip](https://github.com/strimzi/strimzi-kafka-operator/files/6317438/strimzi-kafka-ssl-folder.zip)

i have created secret using certificate folder which has been mentioned in above zip file.

**Reply by scholzj** [reply]
Please share the steps (with the commands you are executing etc.). Even if I try to use the files you shared and they work, it will still not reveal what did you do wrong.

**Reply by Cshanmukh80** [reply]
**before producing commands i am copying the keystore and trustore file manually into the pod by using following commands**

kubectl cp kafka.keystore.jks kafka-kafka-0:/tmp/kafka/
kubectl cp kafka.truststore.jks kafka-kafka-0:/tmp/kafka/

**This is my ssl.properties file before producing the message**

echo 'security.protocol=SSL
ssl.truststore.location=/tmp/kafka/kafka.truststore.jks
ssl.truststore.password=kafka123
ssl.keystore.location=/tmp/kafka/kafka.keystore.jks
ssl.keystore.password=kafka123
ssl.enabled.protocols=TLSv1.2
ssl.endpoint.identification.algorithm=' > clientt-ssl.properties

before producing i am creating the above ssl.properties file inside the pod

**kafka producing command i am using**

bin/kafka-console-producer.sh --broker-list localhost:9093 --topic test --producer.config client-ssl.properties


these steps i am doing everytime

**Reply by scholzj** [reply]
So what does the TLS debug log tell you?

**Reply by scholzj** [reply]
Also, you still did not shared how did you created the truststore and keystore.

**Reply by Cshanmukh80** [reply]
**The below steps i am using to generate certificates for my kafka**
- 1) Generate certificate for each broker kafka:
COMANDO: keytool -keystore server.keystore.jks -alias localhost -validity 365 -genkey

- 2) Create CA. The generated CA is a public-private key pair and certificate used to sign other certificates. A CA is responsible for signing certificates. 
COMANDO: openssl req -new -x509 -keyout ca-key -out ca-cert -days 365

- 3) Sign all brokers certificates with the generated CA
Export the certificate from the keystore: keytool -keystore server.keystore.jks -alias localhost -certreq -file cert-file
Sign it with the CA: openssl x509 -req -CA ca-cert -CAkey ca-key -in cert-file -out cert-signed -days {validity} -CAcreateserial -passin pass:{ca-password}

- 4) Import both the certificate of the CA and the signed certificate into the keystore:
keytool -keystore server.keystore.jks -alias CARoot -import -file ca-cert
keytool -keystore server.keystore.jks -alias localhost -import -file cert-signed

- 5) Import CA to client truststore and broker/server truststore:
keytool -keystore server.truststore.jks -alias CARoot -import -file ca-cert
keytool -keystore client.truststore.jks -alias CARoot -import -file ca-cert

**Reply by scholzj** [reply]
So what from this ends up in the secret to you configured as the listener certificate? You are really wasting people's time by not providing all the steps as you were asked and providing just some part which you think is sufficient. 

Also, why are you trying to use your custom certificates when you self-sign them anyway? Why don't you use what Strimzi has working out of the box? 

In general, assuming the signed certificate from step 3 is what you have in the secret later and configured in then the truststore should be good. But it cannot be used as a keystore. For keystore, you have to create a user using the `KafkaUser` resource => that creates a secret with the keystore certificates which you should use.

**Reply by Cshanmukh80** [reply]
sorry sir, i didn't create create those certificates which i got from senior employee and also i am a fresher. But i can provide some information why the handshake issue is coming. Please have a look below

![Screenshot from 2021-04-16 17-44-51](https://user-images.githubusercontent.com/80450247/115023843-ef512b80-9edc-11eb-8cb8-887c4fa0ddba.png)
![Screenshot from 2021-04-16 17-45-36](https://user-images.githubusercontent.com/80450247/115023862-f37d4900-9edc-11eb-893d-9484b23bf8f5.png)
![Screenshot from 2021-04-16 17-46-13](https://user-images.githubusercontent.com/80450247/115023917-03952880-9edd-11eb-89bf-90cb92918da6.png)
![Screenshot from 2021-04-16 17-46-39](https://user-images.githubusercontent.com/80450247/115023957-127bdb00-9edd-11eb-92a7-db910f6edfc6.png)

please let me know what are these issues

**Reply by scholzj** [reply]
Sorry, you would really need to share the full log as a text file. It someone else created the certificate then your truststore would not match the server one. The client cannot have any certificates. It needs to have the right certiicate. the truststore in your client needs to match the server cert in the broker. And the keystore in your client needs to match the Clistens CA in the broker.

**Reply by Cshanmukh80** [reply]
[handshake_issue_logs.txt](https://github.com/strimzi/strimzi-kafka-operator/files/6325095/handshake_issue_logs.txt)

**Reply by scholzj** [reply]
You have it either badly configured or the log is still not complete. Normally on the beginning it should log the content of the keystore and truststore. If it doesn't, it is probably somehow badly configured.

**Reply by scholzj** [reply]
But as I said before. It looks like your keystore is not signed by the Clients CA. Did you use the KafkaUser custom resource to create it as I mentioned before?
